### PR TITLE
FIX Restore LoRA hotswapping functionality

### PR DESCRIPTION
**Pull request by BenjaminBossan**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=45682)

# What does this PR do?

LoRA hotswapping was added in #41297. Due to changes in #43261, it stopped working. This PR restores the functionality.

The tests already cover this and are failing, but probably no one noticed because they're slow tests. On main, they fail with mismatched sizes, which is expected as the padding of the LoRA weights is not being applied. With this PR, I can confirm that the tests pass locally.

Since the two PRs were released in together in v5, there was never a Transformers release with working hotswapping functionality.

### Notes

The hotswap path does not use `_load_pretrained_model`, which means that loading the `state_dict` if not present is required. I hoisted that functionality from the TP path, which was already there, to re-use the same logic. I also apply weight renamings for that reason.

Moreover, I moved the inference model logic to a local function, again to avoid duplicating the logic.

## Code Agent Policy

- [x] I confirm that this is not a pure code agent PR. Claude was used to assist with writing this PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
> 1. Do you know whether text encoder hotswapping is something that’s planned, either in Transformers/PEFT or via a corresponding diffusers change?

If this is merged, it should work directly in Transformers.

> 2\. IIs there a workaround I can use for LoRAs (SDXL) that include text encoder weights?

Hotswapping is already possible in PEFT itself: https://huggingface.co/docs/peft/v0.19.0/en/package_reference/hotswap. This PR is purely about providing the functionality directly in the PEFT-Transformers integration.

**Comment by BenjaminBossan** [comment]
> I can open a new issue or discussion. I don't want to pollute this PR with unrelated comments.

No need, we are aware and we need to enable hotswapping in Transformers first, as Diffusers uses Transformers for the text models.

File: src/transformers/integrations/peft.py [hunk 1]
```diff
@@ -618,45 +620,92 @@ def load_adapter(
...
-        # is not compatible with the way PEFT adapter should be sharded.
-        has_tp_adapters = False
-        for module in self.modules():
-            tp_info = getattr(module, "_tp_info", None)
-            if tp_info is not None:
-                has_tp_adapters = True
-                break
-
-        if has_tp_adapters:
+        def _resolve_adapter_state_dict():
```
Comment (ArthurZucker) [inline src/transformers/integrations/peft.py]: let's not have nested function either out or just don't have a func please.

File: src/transformers/integrations/peft.py [hunk 2]
```diff
@@ -618,45 +620,92 @@ def load_adapter(
...
                 merged_state_dict = {}
                 for ckpt_file in checkpoint_files:
                     merged_state_dict.update(load_state_dict(ckpt_file))
-            else:
-                raise ValueError("Neither a state dict nor checkpoint files were found.")
+                return merged_state_dict
+            raise ValueError("Neither a state dict nor checkpoint files were found.")
 
-            adapter_state_dict = merged_state_dict
+        def set_inference_mode(model):
```
Comment (ArthurZucker) [inline src/transformers/integrations/peft.py]: same

File: src/transformers/integrations/peft.py [hunk 3]
```diff
@@ -618,45 +620,92 @@ def load_adapter(
...
+            model.eval()
+            for module in model.modules():
+                if isinstance(module, BaseTunerLayer):
+                    module.requires_grad_(False)
+
+        # If the model is tensor parallel, we handle the sharding of the state dict here since the logic in `self._load_pretrained_model`
+        # is not compatible with the way PEFT adapter should be sharded.
```
Comment (ArthurZucker) [inline src/transformers/integrations/peft.py]: can you add details please? Also this is not expected and does not seem "true" as I don't see any sharding done

**Comment by ArthurZucker** [comment]
Also my main comment is that this code should be inside peft directly and the bridge layer should be as small as possible.

File: src/transformers/integrations/peft.py (hunk 3 above)
Comment (BenjaminBossan) [reply src/transformers/integrations/peft.py]: Note that this is not my comment: https://github.com/huggingface/transformers/blob/v5.8.0/src/transformers/integrations/peft.py#L621-L622

File: src/transformers/integrations/peft.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/transformers/integrations/peft.py]: Moved to a separate method.

File: src/transformers/integrations/peft.py (hunk 2 above)
Comment (BenjaminBossan) [reply src/transformers/integrations/peft.py]: Moved to a separate method.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the review @ArthurZucker, I updated the PR according to your feedback.

> Also my main comment is that this code should be inside peft directly and the bridge layer should be as small as possible.

In general, I also prefer if users do `peft_model.load_adapter` etc. but at some point, it was decided it would be good to have direct transformers integration and thus `transformers_model.load_adapter` etc. were added. I think there is some utility there but we can think about reducing the amount of code required for that. But let's handle that in a separate PR.

File: src/transformers/integrations/peft.py [hunk 4]
```diff
@@ -428,6 +429,45 @@ class PeftAdapterMixin:
     _prepare_peft_hotswap_kwargs: dict | None = None
     peft_config: dict[str, PeftConfigLike]
 
+    def _resolve_adapter_state_dict(
```
Comment (BenjaminBossan) [inline src/transformers/integrations/peft.py:87]: Note: This is not new code, it's just moved to a separate method to avoid duplication. Original code here: https://github.com/huggingface/transformers/blob/049d2bf1220747b6d39e2a978b9f5fe0defa1dca/src/transformers/integrations/peft.py#L631-L656

File: src/transformers/integrations/peft.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/integrations/peft.py]: its a good occasion to remove it then 😉

File: src/transformers/integrations/peft.py [hunk 5]
```diff
@@ -628,35 +668,54 @@ def load_adapter(
...
+            # need to apply conversions manually as we don't use _load_pretrained_model
+            renamings = [r for r in peft_weight_conversions if isinstance(r, WeightRenaming)]
+            converters = [c for c in peft_weight_conversions if isinstance(c, WeightConverter)]
+            meta_state_dict = self.state_dict()
+            processed_state_dict = {}
+            for key, value in adapter_state_dict.items():
+                renamed_key, _ = rename_source_key(key, renamings, converters, self.base_model_prefix, meta_state_dict)
+                processed_state_dict[renamed_key] = value
+
+            check_hotswap_configs_compatible(self.peft_config[adapter_name], peft_config)
```
Comment (ArthurZucker) [inline src/transformers/integrations/peft.py:384]: we are renaming but we are not running the weight conversion operations tho

**Review by ArthurZucker — COMMENTED** [review]
Can you add a test with a mixtral model please! Bypassing core model loading is not great in general!

File: src/transformers/integrations/peft.py (hunk 3 above)
Comment (BenjaminBossan) [reply src/transformers/integrations/peft.py]: Done

File: src/transformers/integrations/peft.py (hunk 5 above)
Comment (BenjaminBossan) [reply src/transformers/integrations/peft.py:384]: From reading `rename_source_key`, it appears like the `converters` are being applied there. Is that incorrect?

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the review, your comments should be addressed. I added mixtral tests too.

**Comment by BenjaminBossan** [comment]
AFAICT, failing CI is unrelated to this PR.

**Comment by BenjaminBossan** [comment]
@ArthurZucker I merged the latest main and resolved a merge conflict. The failing CI seems to be unrelated:

- `tests/models/glm_moe_dsa/test_modeling_glm_moe_dsa.py::GlmMoeDsaModelTest::test_tp_backward`
- `tests/models/olmo3/test_modeling_olmo3.py::Olmo3ModelTest::test_tp_generation_quantized`
- `tests/models/mistral/test_modeling_mistral.py::MistralModelTest::test_tp_generation`

File: src/transformers/integrations/peft.py (hunk 5 above)
Comment (ArthurZucker) [reply src/transformers/integrations/peft.py:384]: nope! converters.convert() is what you need :)

Comment (ArthurZucker) [reply src/transformers/integrations/peft.py:384]: https://github.com/huggingface/transformers/blob/a8f150d35d5863971db1e5c1dbc2a1c265f27f96/src/transformers/core_model_loading.py#L1559-L1568 or https://github.com/huggingface/transformers/blob/a8f150d35d5863971db1e5c1dbc2a1c265f27f96/src/transformers/core_model_loading.py#L1461-L1467

**Review by ArthurZucker — COMMENTED** [review]
I want to merge but this does not work if you swap mixtral for mixtral -> its missing that test with non v5 ckpt

**Comment by BenjaminBossan** [comment]
Thanks for the review @ArthurZucker. I indeed misunderstood `rename_source_key` and now the weight conversion logic should be properly reimplemented for the hotswap path. I also found an issue with the mixtral test not working with `torch.compile(..., mode="reduce-overhead")` that I fixed.

> but this does not work if you swap mixtral for mixtral -> its missing that test with non v5 ckpt

Could you please share how you tested that? Hotswapping currently does not work with MoE layers (as they require targeting the `nn.Parameter` directly, which needs to be added in PEFT) so I think this is not working no matter what.

Failing CI seems to be unrelated:

- `tests/models/kosmos2_5/test_modeling_kosmos2_5.py::Kosmos2_5ModelTest::test_generate_with_cache_matches_no_cache`
- `tests/models/gpt_oss/test_modeling_gpt_oss.py::GptOssModelTest::test_ep_forward`

**Comment by ArthurZucker** [comment]
I did not test I was more saying if you have any weight conversion ops -> mixtral does and is a simple case! 
Ok having a look

**Comment by BenjaminBossan** [comment]
> I did not test I was more saying if you have any weight conversion ops -> mixtral does and is a simple case!

Ah okay, but I think even with correct conversion, it would not work for the reason mentioned above. I think it's enough of an edge case (weight conversion + hotswap + no MoE) that we can ignore it for now.

File: tests/peft_integration/test_peft_integration.py [hunk 6]
```diff
@@ -1313,3 +1327,17 @@ def test_maybe_load_adapters_path_not_overwritten_for_complete_model(self):
             # Load from the saved path and make sure it actually loads despite
             # the invalid adapter config path
             AutoModel.from_pretrained(tmp_dir)
+
+    def test_mixtral_hotswap_without_compile_works(self):
+        # test a model that usees weight conversion
+        model_id = "hf-internal-testing/Mixtral-tiny"
```
Comment (ArthurZucker) [inline tests/peft_integration/test_peft_integration.py]: are the experts properly targeted ? (the only that actually do `convert`?

**Review by ArthurZucker — APPROVED** [review]
Ty!

File: tests/peft_integration/test_peft_integration.py (hunk 6 above)
Comment (BenjaminBossan) [reply tests/peft_integration/test_peft_integration.py]: No, the experts are not covered by this test. As mentioned, it wouldn't work for lack of support in PEFT itself. I have an item on my TODO list to see if expert (or rather, `nn.Parameter`) targeting can be supported with hotswapping. When/if that lands, I can update the Transformers test to check for that.

**Comment by BenjaminBossan** [comment]
@ArthurZucker I brought the PR up-to-date with the latest Transformers changes. I also tackled the issue of whether hotswapping can work with v4->v5 weight conversion. Turns out that there wasn't really any blocker from the PEFT side, just some minor changes needed on Transformers side, which I added to the PR. New tests cover this case. (Minor note: Targeting `nn.Parameter`s with `torch.compile` enabled *will* trigger re-compilation. But that's not directly related to this PR, it's a fundamental issue with using `nn.utils.parametrize`, which is the PEFT mechanism for targeting `nn.Parameter`s.)

PEFT tests are now part of the Transformers CI and passed, but the hotswap tests are slow tests, so I ran them locally with GPU to confirm that they pass. Also some tests are guarded behind PEFT >= v0.20.0, I commented those guards out and ensured they pass with PEFT from main.

The failing CI tests seem to be unrelated to this PR.

**Comment by BenjaminBossan** [comment]
The PR was synced with the latest main. I checked all failing tests and they all look unrelated to this PR.
